x-pack/plugins/reporting/server/export_types/common/pdf/worker.ts

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Ensure, SerializableRecord } from '@kbn/utility-types';
+
+import { workerData, parentPort, isMainThread } from 'worker_threads';
+import { i18n } from '@kbn/i18n';
+import _ from 'lodash';
+import path from 'path';
+import Printer from 'pdfmake';
+// @ts-ignore: no module definition
+import concat from 'concat-stream';
+
+import { getTemplate } from './get_template';
+import type { TemplateLayout } from './types';
+import { assetPath } from './constants';
+
+export type PdfWorkerData = Ensure<
+  {
+    layout: TemplateLayout;
+    title: string;
+    content: SerializableRecord[];
+
+    logo?: string;
+  },
+  SerializableRecord
+>;
+
+if (!isMainThread) {
+  execute();
+}
+
+async function execute() {
+  const tableBorderWidth = 1;
+
+  const fontPath = (filename: string) => path.resolve(assetPath, 'fonts', filename);
+
+  const fonts = {
+    Roboto: {
+      normal: fontPath('roboto/Roboto-Regular.ttf'),
+      bold: fontPath('roboto/Roboto-Medium.ttf'),
+      italics: fontPath('roboto/Roboto-Italic.ttf'),
+      bolditalics: fontPath('roboto/Roboto-Italic.ttf'),
+    },
+    'noto-cjk': {
+      // Roboto does not support CJK characters, so we'll fall back on this font if we detect them.
+      normal: fontPath('noto/NotoSansCJKtc-Regular.ttf'),
+      bold: fontPath('noto/NotoSansCJKtc-Medium.ttf'),
+      italics: fontPath('noto/NotoSansCJKtc-Regular.ttf'),
+      bolditalics: fontPath('noto/NotoSansCJKtc-Medium.ttf'),
+    },
+  };
+
+  const printer = new Printer(fonts);
+
+  const { layout, logo, title, content } = workerData as PdfWorkerData;
+
+  const docDefinition = _.assign(getTemplate(layout, logo, title, tableBorderWidth, assetPath), {
+    content,
+  });
+
+  const pdfDoc = printer.createPdfKitDocument(docDefinition, {
+    tableLayouts: {
+      noBorder: {
+        // format is function (i, node) { ... };
+        hLineWidth: () => 0,
+        vLineWidth: () => 0,
+        paddingLeft: () => 0,
+        paddingRight: () => 0,
+        paddingTop: () => 0,
+        paddingBottom: () => 0,
+      },
+    },
+  });
+
+  const buffer = await new Promise((resolve, reject) => {
+    if (!pdfDoc) {
+      throw new Error(
+        i18n.translate(
+          'xpack.reporting.exportTypes.printablePdf.documentStreamIsNotgeneratedErrorMessage',
+          {
+            defaultMessage: 'Document stream has not been generated',
+          }
+        )
+      );
+    }
+
+    const concatStream = concat(function (pdfBuffer: Buffer) {
+      resolve(pdfBuffer);
```

Comment:
Just curious: I'm not sure how often `concatStream` gets called, but if we call `parentPort!.postMessage(pdfBuffer);` here, would that stream the data back to the parent thread?

We should aim for a way to stream the PDF data back to the parent thread, and to see if we can [connect it to the ContentStream in the execute_job](https://github.com/elastic/kibana/blob/8.1/x-pack/plugins/reporting/server/export_types/printable_pdf_v2/execute_job.ts#L64).

It's secondary in priority to moving PDF compilation off the main event loop, and might not be feasible if `pdfmake` doesn't support streaming.